File: companies/New Folder With Items/3904.smallest-stable-index-ii.go
package leetcode

//
// @lc app=leetcode id=3904 lang=golang
//
// [3904] Smallest Stable Index II
//
// Notes
// suffixMin[i] stores the minimum value from i to the end, while prefixMax is
// updated left to right. The first index where prefixMax - suffixMin[i] <= k is
// the stable index requested by the Python solution. Time: O(n). Space: O(n).
//
// @lc code=start

func FirstStableIndex3904(nums []int, k int) int {
	n := len(nums)
	suffixMin := make([]int, n)
	suffixMin[n-1] = nums[n-1]
	for i := n - 2; i >= 0; i-- {
		if nums[i] < suffixMin[i+1] {
			suffixMin[i] = nums[i]
		} else {
			suffixMin[i] = suffixMin[i+1]
		}
	}

	prefixMax := 0
	for i, value := range nums {
		if value > prefixMax {
			prefixMax = value
		}
		if prefixMax-suffixMin[i] <= k {
			return i
		}
	}
	return -1
}

// @lc code=end

File: pt-1/3904.smallest-stable-index-ii.java
/*
 * @lc app=leetcode id=3904 lang=java
 *
 * [3904] Smallest Stable Index II
 *
 * Precompute suffix minimums. Scan left-to-right while maintaining prefix max;
 * the first index where prefixMax - suffixMin[i] <= k is stable.
 *
 * Time: O(n). Space: O(n).
 */

// @lc code=start
class Solution {
    public int firstStableIndex(int[] nums, int k) {
        int n = nums.length;
        int[] suffixMin = new int[n];
        suffixMin[n - 1] = nums[n - 1];
        for (int i = n - 2; i >= 0; i--) {
            suffixMin[i] = Math.min(nums[i], suffixMin[i + 1]);
        }

        int prefixMax = 0;
        for (int i = 0; i < n; i++) {
            prefixMax = Math.max(prefixMax, nums[i]);
            if (prefixMax - suffixMin[i] <= k) {
                return i;
            }
        }
        return -1;
    }
}
// @lc code=end
